src/mongocxx/private/client_session.hh

```diff
@@ -77,7 +77,11 @@ class client_session::impl {
         std::unique_ptr<mongoc_session_opt_t, decltype(libmongoc::session_opts_destroy)> opt_t{
             libmongoc::session_opts_new(), libmongoc::session_opts_destroy};
 
-        libmongoc::session_opts_set_causal_consistency(opt_t.get(), _options.causal_consistency());
+	if(_options.causal_consistency())
+         libmongoc::session_opts_set_causal_consistency(opt_t.get(), *_options.causal_consistency());
+
+	if(_options.snapshot())
+         libmongoc::session_opts_set_snapshot(opt_t.get(), *_options.snapshot());
```

Comment:
As mentioned in separate comments, I instead recommend preserving the `bool ()` function interface for `causal_consistency()` and `snapshot()` such that the proposed changes here would instead be:

```suggestion
        libmongoc::session_opts_set_causal_consistency(opt_t.get(), _options.causal_consistency());
        libmongoc::session_opts_set_snapshot(opt_t.get(), _options.snapshot());
```

This would also greatly simplify the corresponding test code which must currently deal with handling "unset" vs. "set" values explicitly.

If instead we are in favor of the `optional<bool> ()` interface and deferring default value logic to the C driver, then the "session option" test in [src/mongocxx/test/options/client_session.cpp:28](https://github.com/mongodb/mongo-cxx-driver/blob/c8054911366f802066443565814f2dc766bee38b/src/mongocxx/test/options/client_session.cpp#L28) must be updated (or removed) accordingly.